### PR TITLE
Support @ClientRegistrationId at Class Level

### PR DESCRIPTION
**Pull request by bbudano**
If `@ClientRegistrationId` is not found on a method, the declaring class will be checked instead.

Closes gh-17806

### CODE REVIEW

**Comment by Inder782** [comment]
Sorry for the confusion! I accidentally referenced the wrong issue above. This PR is not related to that issue.

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/web/client/ClientRegistrationIdProcessor.java [hunk 1]
```diff
@@ -37,17 +38,20 @@ public final class ClientRegistrationIdProcessor implements HttpRequestValues.Pr
 
 	public static ClientRegistrationIdProcessor DEFAULT_INSTANCE = new ClientRegistrationIdProcessor();
 
+	private ClientRegistrationIdProcessor() {
```
Comment (rwinch) [inline oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/web/client/ClientRegistrationIdProcessor.java]: Please don't make changes that are unrelated to the PR

File: oauth2/oauth2-client/src/test/java/org/springframework/security/oauth2/client/web/client/ClientRegistrationIdProcessorTests.java [hunk 2]
```diff
@@ -48,32 +50,42 @@ void processWhenClientRegistrationIdPresentThenSet() {
...
 
 		String registrationId = ClientAttributes.resolveClientRegistrationId(builder.build().getAttributes());
-		assertThat(registrationId).isEqualTo(RestService.REGISTRATION_ID);
+		assertThat(registrationId).isEqualTo(REGISTRATION_ID);
 	}
 
 	@Test
 	void processWhenNoClientRegistrationIdPresentThenNull() {
 		HttpRequestValues.Builder builder = HttpRequestValues.builder();
-		Method hasClientRegistrationId = ReflectionUtils.findMethod(RestService.class, "noClientRegistrationId");
```
Comment (rwinch) [inline oauth2/oauth2-client/src/test/java/org/springframework/security/oauth2/client/web/client/ClientRegistrationIdProcessorTests.java]: While these changes are valuable, please stick to the changes related to this ticket. This helps us to track issues and ensure that, when applicable (e.g. a bug fix that is applicable to older supported branches), we can easily selectively apply changes to other branches.

File: oauth2/oauth2-client/src/test/java/org/springframework/security/oauth2/client/web/client/ClientRegistrationIdProcessorTests.java [hunk 3]
```diff
@@ -86,9 +98,16 @@ interface RestService {
...
 
 	@Retention(RetentionPolicy.RUNTIME)
-	@ClientRegistrationId(RestService.REGISTRATION_ID)
+	@ClientRegistrationId(REGISTRATION_ID)
 	@interface MetaClientRegistrationId {
 
 	}
 
+	@ClientRegistrationId(REGISTRATION_ID)
+	interface AnnotatedRestService {
```
Comment (rwinch) [inline oauth2/oauth2-client/src/test/java/org/springframework/security/oauth2/client/web/client/ClientRegistrationIdProcessorTests.java]:
```suggestion
	interface TypeAnnotatedRestService {
```

This is more clear since all of the interfaces are annotated.

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/web/client/ClientRegistrationIdProcessor.java [hunk 4]
```diff
@@ -37,17 +38,20 @@ public final class ClientRegistrationIdProcessor implements HttpRequestValues.Pr
 
 	public static ClientRegistrationIdProcessor DEFAULT_INSTANCE = new ClientRegistrationIdProcessor();
```
Comment (rwinch) [inline oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/web/client/ClientRegistrationIdProcessor.java:40]:
```suggestion

	private SecurityAnnotationScanner<ClientRegistrationId> scanner = SecurityAnnotationScanners.requireUnique(ClientRegistrationId.class);
```

Use a `SecurityAnnotationScanner<ClientRegistrationId>` instead. This will automatically follow the same logic as we do for method security. It ensures that we do not get duplicate annotations which might provide conflicting information and thus use the wrong registration id.

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/web/client/ClientRegistrationIdProcessor.java [hunk 5]
```diff
@@ -37,17 +38,20 @@ public final class ClientRegistrationIdProcessor implements HttpRequestValues.Pr
...
+	}
+
 	@Override
 	public void process(Method method, MethodParameter[] parameters, @Nullable Object[] arguments,
 			HttpRequestValues.Builder builder) {
-		ClientRegistrationId registeredId = AnnotationUtils.findAnnotation(method, ClientRegistrationId.class);
+		ClientRegistrationId registeredId = Optional
+			.ofNullable(AnnotationUtils.findAnnotation(method, ClientRegistrationId.class))
+			.orElseGet(() -> AnnotationUtils.findAnnotation(method.getDeclaringClass(), ClientRegistrationId.class));
+
```
Comment (rwinch) [inline oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/web/client/ClientRegistrationIdProcessor.java]:
```suggestion
		ClientRegistrationId registeredId = this.scanner.scan(method, method.getDeclaringClass());
```

File: oauth2/oauth2-client/src/test/java/org/springframework/security/oauth2/client/web/client/ClientRegistrationIdProcessorTests.java [hunk 6]
```diff
@@ -39,6 +39,8 @@
  */
 class ClientRegistrationIdProcessorTests {
```
Comment (rwinch) [inline oauth2/oauth2-client/src/test/java/org/springframework/security/oauth2/client/web/client/ClientRegistrationIdProcessorTests.java:42]: Add tests that verify failure in the event that there are duplicate annotations. This should be handled by the `SecurityAnnotationScanners.requireUnique`. An example would be `AggregateService.toTest()` should fail due to duplicate annotations at the same level.

```
@ClientRegistrationId("a")
interface AService {}

@ClientRegistrationId("b")
interface BService {}

interface AggregateService extends AService, BService {
	void toTest();
}
```

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/web/client/ClientRegistrationIdProcessor.java (hunk 1 above)
Comment (bbudano) [reply oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/web/client/ClientRegistrationIdProcessor.java]: This change occurred while formatting and wasn't intentional , I will revert it.

File: oauth2/oauth2-client/src/test/java/org/springframework/security/oauth2/client/web/client/ClientRegistrationIdProcessorTests.java (hunk 2 above)
Comment (bbudano) [reply oauth2/oauth2-client/src/test/java/org/springframework/security/oauth2/client/web/client/ClientRegistrationIdProcessorTests.java]: I will revert these changes.

**Comment by rwinch** [comment]
Thanks for the PR! I pushed an example of using `@ClientRegistrationId` at the type level. Once the build passes, this will automatically be merged into main

**Comment by bbudano** [comment]
@rwinch Cool, thanks for the opportunity!
